### PR TITLE
TBOX-102-custom-dataframe-validation

### PR DESCRIPTION
**Pull request by DIvkov575**
# ↪️ Pull Request

## ✔️ PR Todo

- [x] Add documentation
- [x] Add examples
- [x] Ensure you are following the practices in the [contributor guide](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit) and [coding_standards](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards)
  * [Use consistent and descriptive name conventions](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.477s36w4rds9) 
  * [Uses google-style docstrings](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Google-style-docstrings)
  * [Uses versioned API/client methods only](https://docs.google.com/document/d/1c_hhpDMhO0zN793Z_HFOz42Q4g7l1nPRVQ5pXOUsMJU/edit#heading=h.yv6df99fyrq2)
  * [Avoid global variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-global-variables)
  * [Avoid giant try/catch](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Avoid-giant-try-/-except)
  * [Don't use mutable default variables](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Don%E2%80%99t-use-mutable-default-arguments)
  * [Use a `main()` function in your scripts](https://datatamr.atlassian.net/wiki/spaces/FEKB/blog/2020/04/14/1137606830/Customer+Success+Coding+Standards#Use-a-%E2%80%9Cmain()%E2%80%9D--function-in-your-scripts)
- [x] Added/updated testing for this change (ensure you are testing any changes/new code!)
- [x] Update title to link to TBOX issue (i.e. start pull request title with TBOX-xx) 
- [x] Approval from first code-reviewer
- [x] Approval from second code-reviewer (only required if a second reviewer has left comments)
- [x] Pass all checks

### CODE REVIEW

**Comment by DIvkov575** [comment]
I accidentally included an old commit. Tried to fix it.
The conflicting file is the same as in develop.

**Comment by skalish** [comment]
> I accidentally included an old commit. Tried to fix it. The conflicting file is the same as in develop.

One way of removing that commit is to run `git rebase -i upstream/develop`.  This should bring up a list of your commits.  You can drop the one you accidentally including by changing the word in front of it from `pick` to `drop` or just `d`.

File: tests/data_io/test_validate_dataframe.py
Comment (skalish) [inline tests/data_io/test_validate_dataframe.py]: Can you split this into two statements where you check that both "b" and "c" fail individually?

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]: Can you remove the new comma from the end of this line and other similar ones?

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]: Type-hinting for `check_function` should help indicate that it must return a boolean value.

**Review by skalish — APPROVED** [review]
This looks pretty good. Just a few changes.

**Comment by DIvkov575** [comment]
I cannot view the details for the 3.7-test that failed. Should I push the same code again?

**Comment by clint-richardson** [comment]
This is looking really good! One thing that I think is needed is putting some examples for how to use. This probably means making a `dataframe` folder and an example script here: https://github.com/Datatamer/tamr-toolbox/tree/develop/examples/scripts/data_io

and then also the relevant docs changes:  https://github.com/Datatamer/tamr-toolbox/tree/develop/doc_src/examples

File: tamr_toolbox/utils/client.py
Comment (skalish) [inline tamr_toolbox/utils/client.py]:
```suggestion
    client: Client, api_endpoint: str, *, timeout_seconds: int = 600, sleep_seconds: int = 20
```

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]:
```suggestion
    custom_check_columns: Optional[List[str]] = None,
```

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]:
```suggestion
    df_profile: "pandas.DataFrame", *, require_nonnull_columns: Optional[List[str]] = None
```

**Review by skalish — CHANGES_REQUESTED** [review]
This needs a few small changes related to type-hints and formatting.  @clint-richardson's request for an example is also a very good idea, so please include that too.

**Comment by DIvkov575** [comment]
Tried adding examples. Please let me know if I did so correctly or if I need to add anything.

File: examples/scripts/data_io/dataframe/validate_dataframe.py
Comment (skalish) [inline examples/scripts/data_io/dataframe/validate_dataframe.py]: Can you name this function something more descriptive like `ensure_not_2`?

File: examples/snippets/data_io/dataframe/validate_dataframe.py
Comment (skalish) [inline examples/snippets/data_io/dataframe/validate_dataframe.py]: See the above comment about replacing `"function_name"` with something that looks like a real function.  Also it looks like there is an extra space here that may cause linting issues.
```suggestion
```

File: doc_src/examples/data_io.rst
Comment (skalish) [inline doc_src/examples/data_io.rst]: Looks like a small typo
```suggestion
Data Input/Output
```

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py:323]: I think this is just
```suggestion
    check_function: Callable[Any, bool],
```

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]: This new argument needs to be added to the "Args" section of the docstring.  Make sure it is clear what is being done here.  It looks like this argument actually needs to be a `Tuple[Callable[Any, bool], List[str]]`, meaning you require a function to be passed, along with a list of columns for it to be applied to.  If any value in the passed columns fails the function, then the validation will fail.  This also means the parameter name `custom_check_columns` might be better changed since it isn't just a list of columns.

An alternative would be to split this into two parameters, `custom_check` and `custom_check_columns`, which would just require making sure if one is passed the other is too.

File: examples/snippets/data_io/dataframe/validate_dataframe.py
Comment (skalish) [inline examples/snippets/data_io/dataframe/validate_dataframe.py]: This should still look like run-able code, so you should pass a real function rather than a string.  `my_custom_function` would be just fine.

**Review by skalish — CHANGES_REQUESTED** [review]
Thanks for the continued work on this.  There are a few things to adjust.  I'm also curious: do we want users to be able to do more than one custom check at a time?  I'm not sure how this would change the design but it is worth considering.

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]: Slight wording change to suggest the "tuple"
```suggestion
        custom_check_columns: tuple of custom function and list of columns to apply function to
```

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [reply tamr_toolbox/data_io/dataframe.py:323]: I was wrong here and that's why the CI is failing. Sorry! You were correct, the first "arg" to `Callable` should be `[Any]` (a list).

**Comment by DIvkov575** [comment]
Removing the brackets around 'Any' raised type error, because Callable needs a list of parameters.

**Comment by DIvkov575** [comment]
Not sure if there is anything else to add. About adding multiple custom checks, I talked to Clint when creating the PR, and he decided to just do one at a time.

**Review by skalish — APPROVED** [review]

**Comment by clint-richardson** [comment]
regarding the multiple checks: how hard would it be to take what is here now and just make a list, then iterate? @skalish , or someone else I'll defer to you about design decision here, but yea the original though was let's see what it looks like making one of them work. Then the question is do we want users to call validate once with multiple things passed in, or just multiple times. Considering we already do multiple validations inside of `dataframe.validate` the former feels better to me, but I'll let the current code owners decide (and also decide if it needs to go in this PR or if we can later make it a tuple or a list of tuples)

File: examples/scripts/data_io/dataframe/validate_dataframe.py
Comment (clint-richardson) [inline examples/scripts/data_io/dataframe/validate_dataframe.py]: do we want an example where we have nested function definitions?

**Comment by DIvkov575** [comment]
Ok I will add it. something like this [[func, cols()], [func, cols()]]

File: examples/scripts/data_io/dataframe/validate_dataframe.py
Comment (skalish) [inline examples/scripts/data_io/dataframe/validate_dataframe.py]: Make sure the variable change gets made in the examples:
```suggestion
        custom_check=[ensure_not_2, ["a", "b"]],
```

File: examples/snippets/data_io/dataframe/validate_dataframe.py
Comment (skalish) [inline examples/snippets/data_io/dataframe/validate_dataframe.py]:
```suggestion
    custom_check=[my_custom_function, ["column_1", "column_2"]],
```

File: examples/snippets/data_io/dataframe/validate_dataframe.py
Comment (skalish) [inline examples/snippets/data_io/dataframe/validate_dataframe.py]:
```suggestion
    custom_check=[my_custom_function, ["column_1", "column_2"]],
```

File: examples/scripts/data_io/dataframe/validate_dataframe.py
Comment (skalish) [reply examples/scripts/data_io/dataframe/validate_dataframe.py]: Do you mean a check function that checks multiple conditions with multiple functions?

Comment (clint-richardson) [reply examples/scripts/data_io/dataframe/validate_dataframe.py]: I just mean I worry about giving people sharp scissors in examples. To me this is slightly sharper scissor than just:

```
def ensure_not_2(value):
....

def main()
```

**Comment by skalish** [comment]
> Ok I will add it. something like this [[func, cols()], [func, cols()]]

This is also what I think is best.  The only downside is that it's a fairly bulky argument, so I recommend making sure the type-hinting and examples in types are `List[Tuple[Callable[[Any], bool], List[str]]]`, e.g. `[(my_func_1, [col_a, col_b]), (my_func_2, [col_a, col_c])]`.

File: examples/scripts/data_io/dataframe/validate_dataframe.py
Comment (clint-richardson) [reply examples/scripts/data_io/dataframe/validate_dataframe.py]: also has the benefit of hinting that you can use any function you import from anywhere

Comment (skalish) [reply examples/scripts/data_io/dataframe/validate_dataframe.py]: Ah, yeah.  It's probably better to have the function defined outside of `main`, so there are no functions defined inside of functions.

**Comment by DIvkov575** [comment]
Is there any reason to do List[Tuple[]] instead of Tuple[Tuple[]]]. The code doesn't try to edit anything passed into the arguments.

**Comment by DIvkov575** [comment]
Added the option to pass in multiple different custom-checks. Please let me know how it looks.

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]: It's not very Pythonic to iterate using range, better to do
```suggestion
        for custom_check in custom_checks:
            failed_checks_dict.update(
                _check_custom(
                    df, check_function=custom_check[0], columns_to_check=custom_check[1],
                ).details
            )
```

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]:
```suggestion
    custom_checks: Tuple[
```
Since this can be multiple, having a plural variable name is a good idea. Don't forget the docstring.

**Comment by skalish** [comment]
> Is there any reason to do List[Tuple[]] instead of Tuple[Tuple[]]]. The code doesn't try to edit anything passed into the arguments.

That's a really good question.  Tuples are fine. This actually gets into some `typing` details. Based on your code, as long as the user passes something we can iterate over (which includes lists, tuples, sets, generators) all will be good.  The `typing` object `Iterable` would be a good generic to use here.

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]:
```suggestion
        custom_checks: collection of tuples each containing a custom function and list of columns on which to apply it
```

File: examples/scripts/data_io/dataframe/validate_dataframe.py
Comment (skalish) [reply examples/scripts/data_io/dataframe/validate_dataframe.py]: @DIvkov575 Can you move this function outside of `main` here?  It's a minor change but I think will make the example clearer to users.

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]: It looks like this got a little confusing, the innermost part should still be a Tuple, just the outermost is a generic Iterable
```suggestion
    custom_checks: Iterable[Callable[[Any], bool], List[str]] = None,
```

Comment (skalish) [reply tamr_toolbox/data_io/dataframe.py]: Now the nice thing about having `custom_checks` be an iterable, is that its default value can actually just be an empty tuple (expressed as `tuple()` or just `()`).  The logic in your function body can then be even simpler, as you don't need to check that `custom_checks is not None`, the `for`-loop will just do nothing if the number of items to loop over is 0.  Does that make sense?

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]: Tests are failing because I erroneously dropped the `Tuple` part in that earlier suggestion:
```suggestion
    custom_checks: Iterable[Tuple[Callable[[Any], bool], List[str]]] = None,
```

File: tests/data_io/test_validate_dataframe.py
Comment (skalish) [inline tests/data_io/test_validate_dataframe.py]:
```suggestion
        return value != 2
```

File: tests/data_io/test_validate_dataframe.py
Comment (skalish) [inline tests/data_io/test_validate_dataframe.py]:
```suggestion
        return value == 3
```

File: examples/scripts/data_io/dataframe/validate_dataframe.py
Comment (skalish) [inline examples/scripts/data_io/dataframe/validate_dataframe.py]:
```suggestion
    return value != 2
```

File: examples/scripts/data_io/dataframe/validate_dataframe.py
Comment (skalish) [inline examples/scripts/data_io/dataframe/validate_dataframe.py]:
```suggestion
    return value == 3
```

File: examples/scripts/data_io/dataframe/validate_dataframe.py
Comment (skalish) [inline examples/scripts/data_io/dataframe/validate_dataframe.py]:
```suggestion
    """Checks the columns of a DataFrame
    
    Returns: ValidationCheck object consisting of Pass/Fail and dict of failed columns
    """
```

File: examples/scripts/data_io/dataframe/validate_dataframe.py
Comment (skalish) [inline examples/scripts/data_io/dataframe/validate_dataframe.py]: Our example scripts should a brief docstring at the top.

Also, it's a nice practice to organize imports.  One way is separating standard library imports (you don't have any of these), third-party imports (`pandas`), and local package imports (`tamr_toolbox`):
```suggestion
"""Example script for validating the contents of a DataFrame"""
import pandas as pd

from tamr_toolbox.data_io import dataframe
from tamr_toolbox.models.validation_check import ValidationCheck
```

**Review by skalish — CHANGES_REQUESTED** [review]
This is very good.  Here are some style changes that should be made before merging, so our examples push users to write code well themselves.  These won't make the code functionally different.

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]:
```suggestion
    custom_checks: Iterable[Tuple[Callable[[Any], bool], List[str]]] = tuple(),
```
What do you think of making this an empty tuple?  `()` is also a valid way of writing it but this might be clearer.

File: tamr_toolbox/data_io/dataframe.py
Comment (skalish) [inline tamr_toolbox/data_io/dataframe.py]: If the above default value for `custom_checks` of `tuple()` (or `()`) is made, this can just be
```suggestion
    for custom_check in custom_checks:
        failed_checks_dict.update(
            _check_custom(
                df, check_function=custom_check[0], columns_to_check=custom_check[1],
            ).details
        )
```

**Review by skalish — COMMENTED** [review]
One more suggestion:

**Review by skalish — APPROVED** [review]
LGTM. Great contribution!
After tests pass, this is ready to merge.  
The linting is unhappy b/c of some whitespace in the example script.  If you run `invoke lint` and `invoke format --fix` you can see, and then autocorrect the issue.

**Comment by DIvkov575** [comment]
I cannot merge due to write access.

**Pull request merged**
